src/analysis/souffle/check_predicate.dl

```diff
@@ -0,0 +1,82 @@
+#ifndef SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+#define SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+
+#include "authorization_logic.dl"
+#include "dataflow_graph.dl"
+#include "tags.dl"
+
+.type Bool <: number
+
+.type CheckPredicate =
+  CP_TagPresence { ap: AccessPath, principal: Principal, tag: Tag }
+  | CP_And { lhs: CheckPredicate, rhs: CheckPredicate }
+  | CP_Or { lhs: CheckPredicate, rhs: CheckPredicate }
+  | CP_Implies { antecedent: CheckPredicate, consequent: CheckPredicate }
+  | CP_Not { to_negate: CheckPredicate }
+
+// A check of some predicate on tags.
+.decl checkP(name: symbol, pred: CheckPredicate)
+.input checkP(delimiter=";")
+
+// A check with an expectation of whether it passes or fails. Checks in the
+// regular `check` relation result in a `checkAndExpectation` with an
+// expectation of PASS.
+.decl checkAndExpectation(
+  name: symbol, pred: CheckPredicate, expectation: symbol)
+.input checkAndExpectation(delimiter=";")
+
+checkAndExpectation(name, pred, "PASS") :- checkP(name, pred).
+
+// Deconstruct all top-level check predicates to get subpredicates here.
+.decl checkSubExprs(pred: CheckPredicate)
+checkSubExprs(pred) :- checkAndExpectation(_, pred, _).
+checkSubExprs(lhs), checkSubExprs(rhs) :- checkSubExprs($CP_And(lhs, rhs)).
+checkSubExprs(lhs), checkSubExprs(rhs) :- checkSubExprs($CP_Or(lhs, rhs)).
+checkSubExprs(ant), checkSubExprs(cons) :- checkSubExprs($CP_Implies(ant, cons)).
+checkSubExprs(inner) :- checkSubExprs($CP_Not(inner)).
+
+// Evaluate predicates based upon their subpredicates. The base case, the
+// `TagPresence` branch, requires looking at the `mayHaveTag` relation, so we
+// define it in taint.dl instead of here.
+.decl predicateEval(pred: CheckPredicate, result: Bool)
+predicateEval($CP_Not(inner), lnot subResult) :-
+  checkSubExprs($CP_Not(inner)), predicateEval(inner, subResult).
+predicateEval($CP_And(lhs, rhs), lhsResult land rhsResult) :-
+  checkSubExprs($CP_And(lhs, rhs)), predicateEval(lhs, lhsResult), predicateEval(rhs, rhsResult).
+predicateEval($CP_Or(lhs, rhs), lhsResult lor rhsResult) :-
+  checkSubExprs($CP_Or(lhs, rhs)), predicateEval(lhs, lhsResult), predicateEval(rhs, rhsResult).
+predicateEval($CP_Implies(ant, cons), (lnot antResult) lor (antResult land consResult)) :-
+  checkSubExprs($CP_Implies(ant, cons)), predicateEval(ant, antResult),
+  predicateEval(cons, consResult).
+
+.decl checkAndResult(name: symbol, result: symbol)
+.output checkAndResult(delimiter=";")
+
+checkAndResult(name, "PASS") :-
+  checkAndExpectation(name, pred, _), predicateEval(pred, result), result != 0.
+
+checkAndResult(name, "FAIL") :-
+  checkAndExpectation(name, pred, _), predicateEval(pred, result), result = 0.
+
+// A check which matches the given expectation.
+.decl checkMatchesExpectation(name: symbol)
+.output checkMatchesExpectation(delimiter=";")
+
+checkMatchesExpectation(name) :- checkAndResult(name, result), checkAndExpectation(name, _, result).
```

Comment:
Should these be part of a test util rather than a part of the core analysis?

Reply:
Perhaps. Let me see if I can refactor to remove test mechanism from the implementation.

Reply:
I've refactored the code. It's not quite what you asked for, as the separation between test utils and regular analysis that we have previously accomplished is sort of an artifact of the approach of generating custom datalog which is compiled per-test.   However, I have made the core analysis code not depend upon test relations and have separated the test relations into a conditionally-compiled, test only portion of the Datalog file.